File: code/java/ch26/practice/Practice02Balanced.java
package ch26.practice;

import java.util.*;

/**
 * Practice 2: Balanced Binary Tree — Chapter 26: Trees — Branches of Logic
 * INSTRUCTIONS: Replace the body of solve() with your solution.
 */
public class Practice02Balanced {
    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int v) { val = v; }
    }

    static TreeNode buildTree(Integer[] values) {
        if (values == null || values.length == 0 || values[0] == null) return null;
        TreeNode root = new TreeNode(values[0]);
        java.util.Queue<TreeNode> queue = new java.util.LinkedList<>();
        queue.add(root);
        int i = 1;
        while (!queue.isEmpty() && i < values.length) {
            TreeNode node = queue.poll();
            if (i < values.length && values[i] != null) {
                node.left = new TreeNode(values[i]);
                queue.add(node.left);
            }
            i++;
            if (i < values.length && values[i] != null) {
                node.right = new TreeNode(values[i]);
                queue.add(node.right);
            }
            i++;
        }
        return root;
    }

    public static boolean solve(TreeNode root) {
        // TODO: Replace this with your solution
        return false;
    }

    public static void main(String[] args) { Scanner sc = new Scanner(System.in); sc.close(); }
}

File: code/python/ch26/practice/practice_02_balanced.py
"""
Practice 2: Balanced Binary Tree
================================
Chapter 26: Trees — Branches of Logic

PROBLEM
-------
Return True if the binary tree is height-balanced.

CONSTRAINTS
-----------
- See test cases for input constraints

HINT
----
Bottom-up: compute height, return -1 if unbalanced. If left or right is -1, or heights differ by > 1, the subtree is unbalanced.

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
"""

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree(values):
    if not values or values[0] is None:
        return None
    root = TreeNode(values[0])
    queue = [root]
    i = 1
    while queue and i < len(values):
        node = queue.pop(0)
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    return root


def solve(root):
    """Return True if the binary tree is height-balanced."""
    pass  # TODO: Replace this with your solution


if __name__ == "__main__":
    tree = build_tree([3, 9, 20, None, None, 15, 7])
    print(solve(tree))
